File: clang/test/Analysis/analyzer-display-progress.m
// RUN: %clang_cc1 -fblocks -analyze -analyzer-display-progress %s 2>&1 | FileCheck %s

#include "Inputs/system-header-simulator-objc.h"

static void f() {}

@interface I: NSObject
-(void)instanceMethod:(int)arg1 with:(int)arg2;
+(void)classMethod;
@end

@implementation I
-(void)instanceMethod:(int)arg1 with:(int)arg2 {}
+(void)classMethod {}
@end

void g(I *i, int x, int y) {
  [I classMethod];
  [i instanceMethod: x with: y];

  void (^block)(void);
  block = ^{};
  block();
}

// CHECK: analyzer-display-progress.m f
// CHECK: analyzer-display-progress.m -[I instanceMethod:with:]
// CHECK: analyzer-display-progress.m +[I classMethod]
// CHECK: analyzer-display-progress.m g
// CHECK: analyzer-display-progress.m block (line: 22, col: 11)

File: clang/test/Analysis/analyzer-display-progress.cpp
// RUN: %clang_analyze_cc1 -analyzer-display-progress %s 2>&1 | FileCheck %s

void f() {};
void g() {};
void h() {}

struct SomeStruct {
  void f() {}
};

struct SomeOtherStruct {
  void f() {}
};

namespace ns {
  struct SomeStruct {
    void f(int) {}
    void f(float, ::SomeStruct) {}
    void f(float, SomeStruct) {}
  };
}

// CHECK: analyzer-display-progress.cpp f()
// CHECK: analyzer-display-progress.cpp g()
// CHECK: analyzer-display-progress.cpp h()
// CHECK: analyzer-display-progress.cpp SomeStruct::f()
// CHECK: analyzer-display-progress.cpp SomeOtherStruct::f()
// CHECK: analyzer-display-progress.cpp ns::SomeStruct::f(int)
// CHECK: analyzer-display-progress.cpp ns::SomeStruct::f(float, ::SomeStruct)
// CHECK: analyzer-display-progress.cpp ns::SomeStruct::f(float, struct ns::SomeStruct)
